### DISCUSSION TITLE
请教Python打包exe的方法

### DISCUSSION THREAD

**Discussion by lon91ong** [SeaHOH/GotoX#225] [问答/咨询]
尝试了流行的Pyinstaller和Nuitka
Pyinstaller打包方便, 成功率高。文件夹打包结果一堆全都放在一个文件夹里面，毫无层次；打包成一个exe不过是自解压套了层壳而已。
Nuitka打包的成功率没有Pyinstaller高，打包出来的东西经常报错，整体打包成一个文件的效果和Pyinstaller一样。

请教GotoX是如何打包的，目录有层次，exe也压缩的足够小巧

**Comment by SeaHOH** [comment]
以下见解仅针对 GotoX，有其它项目需要打包，可参考知乎，有许多优秀的答案和文章。

打包的目的无外乎两点：混淆和便携。混淆我们肯定的不需要的，没啥不可见人的；再说便携，最终目的就是方便随处使用，不一定非要弄成单独一个文件。所以结果就是：

- GotoX 本身完全保留 py 文件，运行时不生成 pyc 文件，目录也根据需要进行设置，以方便用户查看和修改；

- 而环境依赖方面：
    - 针对 Windows 有提供一个特制的便携 Python，包含所有依赖，其中文件尽可能打包成 zip，方便查看；  
       (使用根目录下的 pack_portable 打包，完整流程可参考 .travis.yml 文件，一切都是可见的。)
    - 另有 requirements 文件，方便任意系统包括 Windows 将依赖安装到指定的 Python 环境来使用。

**Reply by SeaHOH** [reply]
最终打包大小主要看压缩方式，和一般文件压缩没有什么区别。我使用的压缩参数可参考 .travis.yml 文件，另外如果便携 Python 中不分别打包单个依赖，保持非压缩状态，整体打包压缩后还能小许多。
